### PR TITLE
WT-2397: Cursor traversal from end of the tree skips records.

### PR DESCRIPTION
**Pull request by keithbostic**
Positioning a cursor at the end of the tree can race with page splits.
During the initial descent of the tree in the cursor.prev case, do the
same tests for races we do for a search past the end of the tree.

### CODE REVIEW

**Comment by keithbostic** [comment]
@agorrod, @michaelcahill: a different approach that works for awhile, but eventually races (using Adam's test program).

**Comment by agorrod** [comment]
@keithbostic This branch passes my local testing, and testing using the test case that was used to report the problem.

File: src/btree/bt_walk.c [hunk 1]
```diff
@@ -163,12 +163,12 @@ __page_ascend(WT_SESSION_IMPL *session,
...
 /*
- * __page_descend --
- *	Descend the tree one level.
+ * __page_descend_prev --
+ *	Descend the tree one level, during a previous-cursor walk.
  */
 static void
-__page_descend(WT_SESSION_IMPL *session,
-    WT_PAGE *page, WT_PAGE_INDEX **pindexp, uint32_t *slotp, bool prev)
+__page_descend_prev(
```
Comment (agorrod) [inline src/btree/bt_walk.c]: For consistency this function should change name to `__ref_descend_prev` due to the parameter change.

File: src/btree/bt_walk.c [hunk 2]
```diff
@@ -233,21 +230,41 @@ __page_descend(WT_SESSION_IMPL *session,
...
 	}
 	*pindexp = pindex;
 }
 
 /*
+ * __page_initial_descent_prev --
+ *	Descend the tree one level, when setting up the initial cursor position
+ * for a previous-cursor walk.
+ */
+static bool
```
Comment (agorrod) [inline src/btree/bt_walk.c]: I'd mark this as inline.

The "initial" in the name makes me think we are descending from the root page, but I see that it matches the local variable in the caller. I don't love it but don't have a good alternative suggestion.

**Comment by agorrod** [comment]
@keithbostic I've made a couple of minor comments. Apart from those I'm happy with this patch.

File: src/btree/bt_walk.c (hunk 1 above)
Comment (keithbostic) [inline src/btree/bt_walk.c]: Agreed, I'll make a consistency pass.

File: src/btree/bt_walk.c (hunk 2 above)
Comment (keithbostic) [inline src/btree/bt_walk.c]: Will do. Yes, the reason I chose the name was to match the local variable in the caller, open to anything better, I'm not excited by the choice either.

**Comment by keithbostic** [comment]
@michaelcahill, can you please review this change?

Alex and I are happy with it, but I'd appreciate another set of eyes.

**Comment by michaelcahill** [comment]
@keithbostic, thanks, I've reviewed and I'm happy with this change.  I'll merge.

**Pull request merged**